src/agent/onefuzz-task/src/tasks/coverage/generic.rs

```diff
@@ -457,49 +457,78 @@ impl<'a> TaskContext<'a> {
         Ok(())
     }
 
-    pub async fn save_and_sync_coverage(&self) -> Result<()> {
-        // JSON binary coverage.
-        let binary = self.coverage.clone();
-        let json = BinaryCoverageJson::V1(BinaryCoverageJsonV1::from(binary));
-        let text = serde_json::to_string(&json).context("serializing binary coverage")?;
-        let path = self.config.coverage.local_path.join(COVERAGE_FILE);
-        fs::write(&path, &text)
-            .await
-            .with_context(|| format!("writing coverage to {}", path.display()))?;
-
-        // JSON source coverage.
-        let source = self.source_coverage().await?;
-        let json = SourceCoverageJson::V1(SourceCoverageJsonV1::from(source.clone()));
-        let text = serde_json::to_string(&json).context("serializing source coverage")?;
-        let path = self.config.coverage.local_path.join(SOURCE_COVERAGE_FILE);
-        fs::write(&path, &text)
-            .await
-            .with_context(|| format!("writing source coverage to {}", path.display()))?;
+    pub async fn save_coverage(
+        coverage: &BinaryCoverage,
+        source_allowlist: &AllowList,
+        binary_coverage_path: &Path,
+        source_coverage_path: &Path,
+        copbertura_file_path: &Path,
+    ) -> Result<()> {
+        Self::save_binary_coverage(coverage, binary_coverage_path)?;
+
+        let source = binary_to_source_coverage(coverage, source_allowlist.clone())?;
+
+        Self::save_source_coverage(&source, source_coverage_path).await?;
+
+        Self::save_cobertura_xml(&source, copbertura_file_path).await?;
+        Ok(())
+    }
 
-        // Cobertura XML source coverage.
-        let cobertura = CoberturaCoverage::from(source.clone());
-        let text = cobertura.to_string()?;
-        let path = self
+    pub async fn save_and_sync_coverage(&self) -> Result<()> {
+        let copbertura_file_path = self
             .config
             .coverage
             .local_path
             .join(COBERTURA_COVERAGE_FILE);
-        fs::write(&path, &text)
-            .await
-            .with_context(|| format!("writing cobertura source coverage to {}", path.display()))?;
 
+        let source_coverage_path = self.config.coverage.local_path.join(SOURCE_COVERAGE_FILE);
+        let binary_coverage_path = self.config.coverage.local_path.join(COVERAGE_FILE);
+
+        Self::save_coverage(
+            &self.coverage,
+            &self.source_allowlist,
+            &binary_coverage_path,
+            &source_coverage_path,
+            &copbertura_file_path,
+        )
+        .await?;
         self.config.coverage.sync_push().await?;
+        Ok(())
+    }
 
+    async fn save_cobertura_xml(source: &SourceCoverage, path: &Path) -> Result<(), anyhow::Error> {
+        let cobertura = CoberturaCoverage::from(source);
+        let cobertura_coverage_file = std::fs::File::create(path)
```

Comment:
We could use the tokio file create here

Reply:
Same for `save_source_coverage` and `save_binary_coverage`

Reply:
this would require a change in `write_xml` to make it work with the tokio buWriter